### PR TITLE
HIVE-25165. Generate & track statistics per event type for incremental load in replication metrics.

### PR DESCRIPTION
**Pull request by ayushtkn**
Same output from replication_metrics:
image

### CODE REVIEW

File: ql/src/java/org/apache/hadoop/hive/ql/exec/repl/ReplStatsTracker.java [hunk 1]
```diff
@@ -0,0 +1,132 @@
...
+
+    // Tracking for top K events, Maintain the list in descending order.
+    ListOrderedMap<Long, Long> topKEntries = topKEvents.get(eventType);
+    if (topKEntries == null) {
+      topKEntries = new ListOrderedMap<>();
+      topKEntries.put(Long.parseLong(eventId), timeTaken);
+      topKEvents.put(eventType, topKEntries);
+    } else {
+      // Get the index of insertion, by descending order.
+      int index = Collections.binarySearch(new ArrayList(topKEntries.values()), timeTaken, Collections.reverseOrder());
```
Comment (aasha) [inline ql/src/java/org/apache/hadoop/hive/ql/exec/repl/ReplStatsTracker.java:72]: where are you sorting this?

File: ql/src/java/org/apache/hadoop/hive/ql/parse/repl/ReplLogger.java [hunk 2]
```diff
@@ -47,4 +48,8 @@ public void dataCopyLog(String message) {
 
   public void setParams(String dbName, String dumpDirectory, long numTables, long numFunctions) {
   }
+
+  public ReplStatsTracker getReplStatsTracker() {
```
[the remark this answers is not in the document]
Comment (ayushtkn) [reply ql/src/java/org/apache/hadoop/hive/ql/parse/repl/ReplLogger.java:53]: Added a NoOpReplStatTracker as suggested

File: ql/src/java/org/apache/hadoop/hive/ql/parse/repl/metric/event/Stage.java [hunk 3]
```diff
@@ -35,6 +35,7 @@
   private Map<String, Metric> metrics = new HashMap<>();
   private String errorLogPath;
   private SnapshotUtils.ReplSnapshotCount replSnapshotCount = new SnapshotUtils.ReplSnapshotCount();
+  private String replStats = "";
```
[the remark this answers is not in the document]
Comment (ayushtkn) [reply ql/src/java/org/apache/hadoop/hive/ql/parse/repl/metric/event/Stage.java]: Done

File: ql/src/java/org/apache/hadoop/hive/ql/exec/repl/ReplStatsTracker.java [hunk 4]
```diff
@@ -0,0 +1,132 @@
...
+/**
+ * Tracks the replication statistics per event type.
+ */
+public class ReplStatsTracker {
+
+  // Maintains the descriptive statistics per event type.
+  private HashMap<String, DescriptiveStatistics> descMap;
+
+  // Maintains the top K costliest eventId's
+  private HashMap<String, ListOrderedMap<Long, Long>> topKEvents;
```
[the remark this answers is not in the document]
Comment (ayushtkn) [reply ql/src/java/org/apache/hadoop/hive/ql/exec/repl/ReplStatsTracker.java]: Changed to ConcurrentHashMap & made addEntry synchronised as well

File: ql/src/java/org/apache/hadoop/hive/ql/exec/repl/ReplStatsTracker.java (hunk 1 above)
Comment (ayushtkn) [reply ql/src/java/org/apache/hadoop/hive/ql/exec/repl/ReplStatsTracker.java:72]: Here itself. :-)
When the first element the list is sorted(only one element), post that this gives the index where to insert so the list stays sorted, so when I insert the element to the index returned from this, the list stays always sorted.

**Pull request merged**